### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by alinalaktionova**

### CODE REVIEW

**Review by didevgen — APPROVED** [review]

File: src/Login/Login.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    const { id, features } = data.authenticate.user;
+    setUserInfo({
+      variables: {
+        key: data.authenticate.token,
+        value: { id: id, features: features }
+      }
+    }).then(res => {
+      if (res.data.setUserInfo === true) {
+        Cookies.set("token", data.authenticate.token, { expires: 1 / 24 });
+        props.history.push("/users");
```
Comment (didevgen) [inline src/Login/Login.tsx]: return <Redirect .../>

File: server/shema/resolvers/mutationResolvers.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      await redis.set(key, JSON.stringify(value));
+      return true;
+    } catch (e) {
+      logger.error(`Failed to set user in redis: ${e.message}`);
+      return false;
+    }
+  },
+
+  async createUser(root, { data }, context) {
+    if (context.currentUser && context.currentUser.features.includes(CREATE)) {
```
Comment (Deadid) [inline server/shema/resolvers/mutationResolvers.js]: I would be nice to separate authentication logic from data persistence layer logic.
Also, seems like we can generalize this somehow

File: server/shema/resolvers/queryResolvers.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  },
+  async getUserInfo(parent, args, context) {
+    try {
+      return await Users.findByPk(context.currentUser.id);
+    } catch (e) {
+      return null;
+    }
+  },
+  async getAllUsers(root, { id }) {
+    return Users.findAll({ where: { id: { [Op.ne]: id }, isDeleted: false } });
```
Comment (Deadid) [inline server/shema/resolvers/queryResolvers.js]: Why we are not using `isActive:true` here?

File: server/features/featureConstants.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+const EDIT = "edit";
+const DELETE = "delete";
+const CREATE = "create";
+
+module.exports = {EDIT, CREATE, DELETE};
```
Comment (Deadid) [inline server/features/featureConstants.js]: Not sure about filename, what about `fetureOperations`?  `Constants` may become a big map which contains basically everything

File: server/shema/resolvers/queryResolvers.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    });
+    if (!user) {
+      return null;
+    }
+    if (await bcrypt.compare(password, user.toJSON().password)) {
+      const token = jwt.sign(
+        { ...JSON.stringify(user) },
+        process.env.AUTH_KEY,
+        {
+          expiresIn: 60 * 60
```
Comment (Deadid) [inline server/shema/resolvers/queryResolvers.js]: Might be worth extracting into constants: `JWT_LIFETIME`.

File: src/App.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+          <RouteHOC
+            path="/users"
+            redirect="/"
+            token={!token}
+            children={
+              <React.Fragment>
+                <ActiveUser>
+                  <CurrentUser />
+                  {data.getUserInfo &&
+                    data.getUserInfo.features.includes("create") && (
```
Comment (Deadid) [inline src/App.tsx]: Might be worth using constant here

File: src/GraphqlOperations/mutationConstants.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import { gql } from "@apollo/client";
```
Comment (Deadid) [inline src/GraphqlOperations/mutationConstants.ts]: I think `constants` in filename is redundant here

File: src/GraphqlOperations/queriesContants.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import { gql } from "@apollo/client";
```
Comment (Deadid) [inline src/GraphqlOperations/queriesContants.ts]: I think `constants` is redundant in filename

File: src/HOC/RouteHOC.tsx [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import React from "react";
+import { Redirect, Route } from "react-router-dom";
+
+const RouteHOC = ({ path, token, redirect, children }: any) => {
```
Comment (Deadid) [inline src/HOC/RouteHOC.tsx]: Not sure whether this really a `HOC`, as far as I know:
`a higher-order component is a function that takes a component and returns a new component.`
This seems to be regular component

File: src/HOC/SettingsCard.tsx [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+interface SettingsCardProps {
+    open: boolean,
+    userId: number,
+    admin?: boolean,
+    initialValues: any,
+    close(): void
+}
+
+const SettingsCard = (props: SettingsCardProps) => {
```
Comment (Deadid) [inline src/HOC/SettingsCard.tsx]: Its not really a HOC, but seems to be regular component

File: server/shema/resolvers/queryResolvers.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        }
+      );
+      return { user: user.toJSON(), token: token };
+    }
+  },
+  async logoutUser(parent, { key }, { redis }) {
```
Comment (Deadid) [inline server/shema/resolvers/queryResolvers.js]: Why this is not mutation? It has side effects

File: server/shema/resolvers/queryResolvers.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
+const Users = require("../../models/UsersModels");
+const { Op } = require("sequelize");
+const jwt = require("jsonwebtoken");
+const bcrypt = require("bcryptjs");
+const logger = require("../../logger/config");
+
+const queries = {
+  async authenticate(root, { login, password }) {
```
Comment (Deadid) [inline server/shema/resolvers/queryResolvers.js]: This seems to be a mutation

File: src/HOC/InputValidateHOC.tsx [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+const InputError = styled.div`
+  margin-bottom: -10px;
+  padding-top: 5px;
+  font-size: 12px;
+  color: #d73c2a;
+  text-align: center;
+`;
+
+const InputValidate: React.FC<Props> = ({ fieldName, id, label, type }) => (
```
Comment (Deadid) [inline src/HOC/InputValidateHOC.tsx]: This is also not a hoc (correct me if I wrong @didevgen )

File: src/ProfileUserInfo/CurrentUser.tsx [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+  const [open, setOpen] = useState(false);
+  if (data.getUserInfo) {
+  }
+  const { name, login, features } = data.getUserInfo;
+
+  const initialValues = {
+    name,
+    login,
+    admin: features.includes("create")
```
Comment (Deadid) [inline src/ProfileUserInfo/CurrentUser.tsx]: seeing this string second time. Also, I'm not sure that this is reliable way to check user role

File: src/Users/PasswordSetup.tsx [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import React, { useState } from "react";
+import { useMutation } from "@apollo/client";
+import { ADD_PASSWORD } from "../GraphqlOperations/mutationConstants";
+import queryString from "query-string";
+import { withRouter, RouteComponentProps } from "react-router";
+
+const PasswordSetup = (props: RouteComponentProps) => {
+  const [password, setPassword] = useState("");
+  const [confirmPassword, setConfirmPassword] = useState("");
+  const [addPassword] = useMutation(ADD_PASSWORD);
```
Comment (Deadid) [inline src/Users/PasswordSetup.tsx:10]: naming proposal: `change password`?

File: src/Users/User.tsx [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+  });
+
+  const closeDialog = () => {
+    setOpen(false);
+  };
+
+  return (
+    <UserItem>
+      <div>Name: {props.name}</div>
+      <div>{props.admin && "admin"}</div>
```
Comment (Deadid) [inline src/Users/User.tsx]: extract all strings to constants please

File: src/Users/User.tsx [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+      {props.getUserInfo && props.getUserInfo.features.includes("create") && (
```
Comment (Deadid) [inline src/Users/User.tsx]: extract all strings to constants please

**Review by Deadid — COMMENTED** [review]
Didn't checkout, so could miss something. 
Also, naming pull requests properly is considered to be a good practice :)

**Pull request merged**